test/test-files/step-function.ts

```diff
@@ -650,3 +650,33 @@ new StepFunction(stack, "fn", async (input: { arr: { name: string }[] }) => {
   input.arr.filter((item) => item.name === "some string");
   input.arr.filter(({ name }) => name === "some string");
 });
+
+/**
+ * Unsupported - Mis-matched element access
+ * @see ErrorCodes.StepFunctions_mismatched_index_type
+ */
+
+new StepFunction(stack, "fn", async () => {
+  const obj = { 1: "value" };
+  const arr = [1];
+
+  // invalid - numeric object property access in SFN is invalid, key must be a string
+  obj[1];
+
+  // invalid - string array access in SFN is invalid, index must be a number
+  arr["0"];
+});
+
+/**
+ * Support - element access
+ */
+
+new StepFunction(stack, "fn", async () => {
+  const obj = { 1: "value" };
+  const arr = [1];
+
+  // valid
+  obj["1"];
+  // valid
+  arr[0];
+});
```

Comment:
What happens when the type is `any`?

Reply:
This would be a very minor edge case as we don't support references

Reply:
A quick test says it fails when the type is `any`, but the error is from an existing validation which checks if the value is a constant.

Reply:
> test/test-files/step-function.ts:667:7 - error Functionless(10025): StepFunctions Invalid collection access